### DISCUSSION TITLE
Consultation on implementating of new model

### DISCUSSION THREAD

**Discussion by llaniewski** [pmgbergen/porepy#1080] [General]
I find the structure of the `porepy` code very promising for implementing some modelling ideas I have. But I think it would save me a lot of time if I could have a talk (zoom call?) with any of the `porepy` developers to ask a couple of questions. Would that be possible?

I would want to develop a model with which I have two (or more) substances simultaneously advected in fractures, with different advection speeds, and would want to consult on the best way forward to do it.

Two additional things, that I did not find how to do (in the documentation) are:
1. how to confine some equations to fractures-only to reduce the number of DOFs and improve performance;
2. how to do a simulation without the matrix and only having 2D fractures someway distributed in 3D space.

**Comment by IvarStefansson** [comment]
Thanks for getting in touch and for your nice words about `porepy`. Could you please elaborate a little bit on the project (target application/purpose of simulation, problem size, etc.) so we can choose the best way to help you? 
To your questions:

0. Advection of  two species is possible. An important point is whether the flow fields are known and constant, or will you solve equations to obtain the flow fields? If so, which?
1. I suggest you look in `momentum_balance.py`, where this is done for the fracture contact mechanics relations. Though this is a different set of equations from what you are interested in, the technique for defining equations on only a few subdomains is the same. 
2. Having only 2D fractures (and their intersections) is only partly supported, but should be possible with some work from your side. Specifically, the FractureNetwork3d class can generate such grids by setting the optional keyword dfn to True (word of caution: such DFN meshes are not much in use, but everything should work.).To create a DFN mesh in the modeling framework, you will need to call the meshing function with this argument from within the model class; this is something we have never done, but it should work.

**Reply by llaniewski** [reply]
Thanks for the quick reply @IvarStefansson. We're doing some research on two-phase and particle-laden flows in narrow fractures at University of Queensland. There are bit different needs for both, but the overarching ideas are similar.

Basically I would want to have two distinct fluxes based on one pressure field and proportion of the two phases (saturation/volume fraction). I see two main roads forward:
- Use the existing single-phase model to track the "mixture", and add another model (similar to energy balance model) for tracking the concentration of one of the phases.
- Make a new model by extending the single-phase model, directly tracking both phases at the same time.
What would you recommend?

There are still some things I don't completely understand about the code. I think the main is the treatment of the aperture in a case with flow and elasticity:
- Where in the code is the "fluid aperture" coupled to the "deformation aperture"? Do I understand correctly that it is coupled in the calculation of `MassBalanceEquations::fluid_mass`, but it's not a AD variable at this point, so the coupling is not accounted in the Jacobian? But if that would be the case, how does the linear solver "know" that changing aperture will change the pressures?

I'll look into the DFN part. I would gladly start with a DFN model before I go into a coupled model.
